### PR TITLE
Training wheels2

### LINKED ISSUE #4404: Unable to save game (training-wheels related?)
After resetting the training wheels progress in the options screen, I started a new random match (player 1, as atlanteans) on 07ba5caa2fcb38a0e7c2ff533fb10699f6856ec4 (language: low german). The story message box teaching how to open the objectives window is currently open. Now, the autosave fails with
```
[00:03:14.817 real] ERROR: GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
[00:03:14.825 real] INFO: SaveHandler::save_game() took 322ms
[00:00:03.174 game] ERROR: Autosave: ERROR! - GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
```
```
[00:04:45.990 real] ERROR: GenericSaveHandler::save_file: data could not be written to file save/wl_autosave_00.wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/wl_autosave_00.wgf/map/scripting/globals.dump
[00:04:45.999 real] INFO: SaveHandler::save_game() took 288ms
[00:01:34.380 game] ERROR: Autosave: ERROR! - GenericSaveHandler::save_file: data could not be written to file save/wl_autosave_00.wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/wl_autosave_00.wgf/map/scripting/globals.dump
```

Manually saving now also fails:
```
[00:05:55.365 real] ERROR: GenericSaveHandler::save_file: data could not be written to file save/testtest.wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/testtest.wgf/map/scripting/globals.dump
```
After opening and closing the objectives window (thus fulfilling the wheel), autosave and manual saving still fail.

---

This only happens on random matches, not in normal games.

### PR DESCRIPTION
**Pull request by gunchleoc**
* New training wheel: Building Spaces
* New training wheel: Flags
* New training wheel: Rocks

* Rework Logs training wheel for treeless maps and player building wrong building
* Fix saveloading for Objectives training wheel
* Improved algorithm for finding buildable field
* Support for indicating dropdown items
* Fix detection of interactive player

Re. https://github.com/widelands/widelands/issues/4396
Fixes https://github.com/widelands/widelands/issues/4404

### CODE REVIEW

**Comment by Noordfrees** [comment]
I got a crash while I was being taught how to use the building spaces on a random map. Possibly related to #4404?
```
[00:00:38.496 real] ERROR: GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
[00:00:38.497 real] INFO: SaveHandler::save_game() took 25ms
[00:00:03.187 game] ERROR: Autosave: ERROR! - GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
[00:00:46.891 real] INFO: Solved training wheel 'objectives'
[00:00:46.891 real] INFO: Running training wheel 'building_spaces.lua'
[00:00:14.576 game] INFO: Gamesave: save requested: Tofällige Kaart (Close this window)
[00:00:14.576 game] INFO: Game: Writing Preload Data ... 
[00:00:14.576 game] INFO: Game: Writing Game Class Data ... 
[00:00:14.576 game] INFO: Game: Writing Player Info ... 
[00:00:14.576 game] INFO: Game: Writing Map Data!
[00:00:14.576 game] INFO: Writing Elemental Data ... 
[00:00:14.576 game] INFO: Writing Player Names And Tribe Data ... 
[00:00:14.576 game] INFO: Writing Port Spaces Data ... 
[00:00:14.576 game] INFO: Writing Heights Data ... 
[00:00:14.576 game] INFO: Writing Terrain Data ... 
[00:00:14.576 game] INFO: Writing Player Start Position Data ... 
[00:00:14.576 game] INFO: Writing Player Message Data ... 
[00:00:14.576 game] INFO: Writing Resources Data ... 
[00:00:14.576 game] INFO: Writing Map Version ... 
[00:00:14.576 game] INFO: Writing Allowed Worker Types Data ... 
[00:00:14.576 game] INFO: Writing Allowed Building Types Data ... 
[00:00:14.576 game] INFO: Writing Flag Data ... 
[00:00:14.576 game] INFO: Writing Road Data ... 
[00:00:14.576 game] INFO: Writing Waterway Data ... 
[00:00:14.576 game] INFO: Writing Building Data ... 
[00:00:14.576 game] INFO: Writing Map Objects ... 
[00:00:14.576 game] INFO: Writing Wincondition Data ... 
[00:00:14.576 game] INFO: Writing Flagdata Data ... 
[00:00:14.576 game] INFO: Writing Buildingdata Data ... 
[00:00:14.576 game] INFO: Writing Node Ownership Data ... 
[00:00:14.576 game] INFO: Writing Players View Data ... 
[00:00:14.576 game] INFO: Writing Scripting Data ... 
PANIC: unprotected error in call to Lua API (Trying to persist a User Interface Panel which is not supported!
stack traceback:)
Abgebrochen (Speicherabzug geschrieben)
```
Version 1.0~git25036[675f4fd@HEAD] (Release)

Edit: This one also happens on official maps, not just on random ones

**Comment by Noordfrees** [comment]
When I start with player number 2 and slot 1 is closed, I get this directly after starting:
```
[00:00:00.010 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/training_wheels/utils/lock.lua"]:7: attempt to index a nil value (local 'player')
```
version 43e6ea1d83cdf2567fae5b4452adac4f69ffb770

**Comment by gunchleoc** [comment]
Thanks for finding those bugs. Should be fixed now.

**Comment by Noordfrees** [comment]
Well, the saveloading bug is still there in Version 1.0~git25049[f772bd5@HEAD] (Debug):
```
[00:00:59.233 real] ERROR: GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
[00:00:59.240 real] INFO: SaveHandler::save_game() took 224ms
[00:00:03.205 game] ERROR: Autosave: ERROR! - GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
```

I completed the Objectives and Building Spaces wheels. Then the output says
```
[00:01:50.494 real] INFO: Closing window 'story_message_box'
[00:01:53.020 real] INFO: Solved training wheel 'building_spaces'
[00:01:53.021 real] INFO: Running training wheel 'flags.lua'
[00:01:53.056 real] INFO: Running training wheel 'logs.lua'
[00:01:53.091 real] INFO: Running training wheel 'rocks.lua'
```
but nothing happens any more…

**Comment by Noordfrees** [comment]
After waiting for 8 minutes gametime (nothing happened) I built a woodcutter's house. As soon as it is complete a window pops up instructing me how to build a – woodcutter's house! I kind of figured this out already…

---

For reference: I was using the random map `8nva-w8d3-k6d3-qds2-07zt-nxcb` 
I was player 2 as frisians, and slot 1 was closed. Win condition: Endless

**Comment by gunchleoc** [comment]
>  As soon as it is complete a window pops up instructing me how to build a – woodcutter's house! I kind of figured this out already…

I originally skipped the woodcutter scenario in that case and marked it as solved, but then we run into problems when the player places the wrong buildings. How about skipping it in the current game only?

Probably nothing happened because it couldn't find a viable building spot. That algorithm can certainly do with more tweaking, so we need to collect information on which map it fails.

**Comment by gunchleoc** [comment]
> > As soon as it is complete a window pops up instructing me how to build a – woodcutter's house! I kind of figured this out already…
> 
> I originally skipped the woodcutter scenario in that case and marked it as solved, but then we run into problems when the player places the wrong buildings. How about skipping it in the current game only?

Done. I still need to dig into the saveloading bug.

**Comment by gunchleoc** [comment]
The saveloading issue is not being caused by 07ba5caa2fcb38a0e7c2ff533fb10699f6856ec4 as I assumed, so I'll dig into that in a separate branch.

**Comment by gunchleoc** [comment]
@Noordfrees are you still getting the saveloading bug? I can't reproduce it. Note that already corrupted savegames won't recover, because the fix for the objectives needed a scripting change.

**Comment by Noordfrees** [comment]
Yes, it still happens with a new random map in 46e9e37@HEAD.
```
[00:00:36.475 real] ERROR: GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
[00:00:36.477 real] INFO: SaveHandler::save_game() took 81ms
[00:00:03.188 game] ERROR: Autosave: ERROR! - GenericSaveHandler::save_file: data could not be written to file save/Tofällige Kaart (Objectives).wgf: RealFSImpl::write: could not open file for writing: /home/benedikt/.widelands/save/Tofällige Kaart (Objectives).wgf/map/scripting/globals.dump
```
To reproduce, just reset the progress, then hit Z in the main menu, and use all default settings and the win condition Endless Game.

**Comment by Noordfrees** [comment]
And perhaps story message windows of multiple wheels shouldn't be allowed to pop up on top of each other? Looks a bit confusing imho:
grafik

**Comment by Niektory** [comment]
> To reproduce, just reset the progress, then hit Z in the main menu, and use all default settings and the win condition Endless Game.

I can reproduce, but only after disabling the "Compress Widelands data files" option.

**Comment by gunchleoc** [comment]
> I can reproduce, but only after disabling the "Compress Widelands data files" option.

Yep, that's it. Now I have to find out what on Earth the difference is between having a zip file or not.

Also, this only happens with the autosave for the objective. All other autosaves are fine.

> And perhaps story message windows of multiple wheels shouldn't be allowed to pop up on top of each other? Looks a bit confusing imho:

This happens because the logs training wheel is solved when the builder arrives, so you don't have to wait for the building to finish before you can continue playing. The only solution here would be to skip the message congratulating the player when the building finishes building. I can certainly remove it if you think that would be better.

**Comment by gunchleoc** [comment]
Autosave bugs should be fixed now - the map scripting packet had some bugs in it that were never triggered before.

**Comment by Noordfrees** [comment]
The autosave works now :) 

After saving manually in the middle of the Logs wheel and reloading I got this:
```
[00:01:17.402 real] INFO: Solved training wheel 'logs'
[00:01:17.403 real] INFO: Solved training wheel 'logs'
[00:00:47.880 game] ERROR: Error in Lua Coroutine
[00:00:47.880 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/training_wheels/utils/buildings.lua..."]:126: attempt to index a nil value (local 'center_field')
[00:00:47.880 game] ERROR: Send message to all players and pause game
[00:00:48.628 game] ERROR: Error in Lua Coroutine
[00:00:48.628 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/training_wheels/utils/buildings.lua..."]:126: attempt to index a nil value (local 'center_field')
[00:00:48.628 game] ERROR: Send message to all players and pause game
```
Version 1.0~git25093[9f328c5@HEAD] (Debug)

**Comment by gunchleoc** [comment]
Thanks! Should be squashed now.

I have also shifted the congratulation message to when the builder arrives and rephrased it a bit. This takes care of the multiple story message boxes.

**Comment by Noordfrees** [comment]
When playing as intended by the wheels everything works fine now. When playing deliberately differently than told by the wheels I get this (probably triggered by the quarry wheel): 
```
[00:07:45.590 game] ERROR: Error in Lua Coroutine
[00:07:45.590 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/training_wheels/flags.lua"]:97: attempt to index a nil value (local 'target_field')
```
Version 1.0~git25100[4b787b1@HEAD] (Debug)

**Comment by gunchleoc** [comment]
I don't see how `wait_for_starting_conditions` can return `nil`. So, I have added another nil check below it.

Not following instructions is an important test; I have done quite a lot of that myself, but of course everybody has different ideas on how to get creative. These scripts turned out to be quite complex to write, so there are probably more bugs to be found down the road.

**Comment by Noordfrees** [comment]
Evil tester removed the constructionsite the moment the builder arrived, and a second later he got this slightly incorrect message:
shot0005
Afterwards, nothing else happens any more – no messages about flags or quarries, even though there are lots of rocks nearby (with a few suited building spaces, just east of the screenshot's view).

Also, when there are no rocks near the starting position, the quarry wheel is never triggered. Not even when I later reach a pile of rocks with many building spaces around.

In my opinion, the fact that the script hides the (initially revealed!) buildhelp only to teach me how to enable it is a bit unexpected and annoying. If the buildhelp is on it should not be turned off.

Version 1.0~git25105[a0dc1e8@HEAD] (Debug)

**Comment by gunchleoc** [comment]
> Evil tester removed the constructionsite the moment the builder arrived

Diligent programmer thwarted the evil tester :)

> Also, when there are no rocks near the starting position, the quarry wheel is never triggered. Not even when I later reach a pile of rocks with many building spaces around.

That's already on my radar. I'll implement that when I create the expansion training wheel, because it will save me time on testing.

> In my opinion, the fact that the script hides the (initially revealed!) buildhelp only to teach me how to enable it is a bit unexpected and annoying. If the buildhelp is on it should not be turned off.

I turned this into a check now, so it will only trigger if the building help is off. I used sleep time to make sure it will start before the logs and rocks if its conditions are met.

**Comment by Noordfrees** [comment]
Evil tester has more tricks in store to foil the diligent programmer's intentions ;)
How am I supposed to follow this piece of advice here?
shot0006

**Comment by gunchleoc** [comment]
I have noted that down in the issue. Fixing this is non-trivial and I'd rather get this branch in if there are no other blockers, because it includes a bugfix.

File: src/scripting/lua_ui.cc [hunk 1]
```diff
@@ -485,6 +497,58 @@ int LuaDropdown::highlight_item(lua_State* L) {
...
+	// Open the dropdown
+	get()->set_list_visibility(true);
+
+	for (; desired_item <= get()->size(); ++desired_item) {
+		y -= get()->lineheight();
+	}
+
+	ipl->set_training_wheel_indicator_pos(Vector2i(x, y));
+
+	return 2;
```
Comment (Noordfrees) [inline src/scripting/lua_ui.cc]:
```suggestion
	return 0;
```

File: src/scripting/lua_map.cc [hunk 2]
```diff
@@ -7113,6 +7129,50 @@ int LuaField::get_buildable(lua_State* L) {
...
+	if (f->get_road(Widelands::WalkingDir::WALK_SW) != Widelands::RoadSegment::kNone) {
+		lua_pushboolean(L, true);
+		return 1;
+	}
+
+	Widelands::FCoords neighbor;
+	const Widelands::Map& map = get_egbase(L).map();
+	map.get_ln(fc, &neighbor);
+	if (neighbor.field->get_road(Widelands::WalkingDir::WALK_E) != Widelands::RoadSegment::kNone) {
+		lua_pushboolean(L, true);
```
Comment (Noordfrees) [inline src/scripting/lua_map.cc:7161]: `return 1;`
same below 2×

File: src/scripting/lua_game.cc [hunk 3]
```diff
@@ -133,6 +136,18 @@ int LuaPlayer::get_name(lua_State* L) {
 	return 1;
 }
 
+/* RST
+   .. attribute:: number
+
+         (RO) The number of this Player on the map.
```
Comment (Noordfrees) [inline src/scripting/lua_game.cc]: A look in the doc would have saved you some typing ;) This is already a property of Player's parent class so you can delete this function again.
https://www.widelands.org/documentation/autogen_wl_bases/#playerbase

File: data/scripting/training_wheels/utils/buildings.lua [hunk 4]
```diff
@@ -40,23 +69,32 @@ end
...
-               return field
+            if sufficient_space then
+               -- Ensure the border won't interfere
+               for nf_idx, nearby_field in ipairs(field:region(2)) do
+                  if player ~= nearby_field.owner then
+                     sufficient_space = false
+                  end
+               end
+               if sufficient_space then
+                     return field
```
Comment (Noordfrees) [inline data/scripting/training_wheels/utils/buildings.lua]: indentation

**Review by Noordfrees — APPROVED** [review]

**Pull request merged**